src/main/scala/uk/gov/nationalarchives/omega/api/services/Dispatcher.scala

```diff
@@ -0,0 +1,107 @@
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package uk.gov.nationalarchives.omega.api.services
+
+import cats.data.Validated.{ Invalid, Valid }
+import cats.data.{ Validated, ValidatedNec }
+import cats.effect.IO
+import cats.effect.std.Queue
+import uk.gov.nationalarchives.omega.api.business._
+import uk.gov.nationalarchives.omega.api.business.echo.{ EchoRequest, EchoResponse, EchoService }
+import uk.gov.nationalarchives.omega.api.services.ServiceIdentifier.ECHO001
+
+import java.util.UUID
+
+class Dispatcher(val localProducer: LocalProducer) {
+
+  def run(dispatcherId: Int)(q: Queue[IO, LocalMessage]): IO[Unit] =
+    for {
+      requestMessage <- q.take
+      _ <- IO.delay {
+             println(s"Dispatcher # $dispatcherId, processing message id: ${requestMessage.persistentMessageId}")
+           }
+      serviceRequest                  <- createServiceRequest(requestMessage)
+      businessService                 <- IO.pure(serviceRequest._1)
+      businessServiceRequest          <- IO.pure(serviceRequest._2)
+      validatedBusinessServiceRequest <- validateBusinessServiceRequest(businessService, businessServiceRequest)
+      businessResult                  <- execBusinessService(businessService, validatedBusinessServiceRequest)
+      res                             <- sendResultToJmsQueue(businessResult, requestMessage)
+    } yield res
+
+  private def createServiceRequest(localMessage: LocalMessage): IO[(BusinessService, BusinessServiceRequest)] =
+    localMessage.serviceId match {
+      case ECHO001 =>
+        IO.delay {
+          Tuple2(
+            new EchoService(),
+            EchoRequest(localMessage.messageText)
+          )
+        }
+      // add more service IDs here
+    }
+
+  private def validateBusinessServiceRequest(
+    businessService: BusinessService,
+    businessServiceRequest: BusinessServiceRequest
+  ): IO[ValidatedNec[RequestValidationError, BusinessServiceRequest]] =
+    IO.delay {
+      businessService match {
+        case value: RequestValidation =>
```

Comment:
When is `businessService` going to be `RequestValidation`? It seems to me it should be an EchoService, always, unless I'm reading this wrong?

Reply:
Looking at EchoService, it extends RequestValidation. I'm not sure why though, I don't understand that inheritance heirachy.